File: LinkedList.cpp
#include <iostream>
using namespace std;

class node
{

private:
	int data;
	node* next;
	node* head;
public:
	node()
	{
		head=NULL;
	}
	
	node* getHead()
	{
		return head; 
	}

	void insertEnd(int item)
	{
		node* temp = new node();
		temp->data=item;
		temp->next=NULL;
		node* temp2 = head;
		if(head==NULL)
			head=temp;
		else
		{				
			while(temp2->next!=NULL)
			{
				temp2=temp2->next;
			}
			temp2->next=temp;
		}

	}

	void insertFront(int item)
	{
		node* temp=new node();
		temp->data=item;
		if(head!=NULL)
			temp->next=head;
		head=temp;
	}

	void insertPos(int item,int pos)
	{
		node* temp = new node();
		temp->data=item;
		int count=1;
		node* temp2 = head;
		while(temp2!=NULL)
		{
			temp2=temp2->next;
			count++;
		}
		if(pos>count)
		{
			cout<<"Invalid position"<<endl;
			return;
		}
		count=1;
		temp2 = head;
		if(head==NULL)
			head=temp;
		else
		{
			while(count<pos-1)
			{
				temp2=temp2->next;
				count++;
			}
			temp->next = temp2->next;
			temp2->next = temp;
		}
		
	}

	void deleteNode(int pos)
	{
		int count = 1;
		node* temp = head;
		if(head==NULL)
		{
			cout<<"Empty List"<<endl;
			return;
		}
		while(temp->next!=NULL)
		{
			temp=temp->next;
			count++;
		}
		if(pos>count)
		{
			cout<<"Invalid position"<<endl;
			return;
		}
		count=1;
		temp=head;
		if(pos == 1)
		{
			head=temp->next;
			delete(temp);
			return;
		}
		while(count<pos-1)
		{
			temp=temp->next;
			count++;
		}
		temp->next=temp->next->next;
	}

	void reverseList()
	{
		node *prev=NULL,*current=head,*next;
		while(current!=NULL)
		{
			next=current->next;
			current->next=prev;
			prev=current;
			current=next;
		}
		head=prev;
	}

	void reverseRecursion(node *temp)
	{
		//temp = head;
		if(temp->next == NULL)
		{
			head = temp;
			return;
		}
		reverseRecursion(temp->next);
		node *temp2 = temp->next;
		temp2->next = temp;
		temp->next = NULL;
	}

	void displayRecursion(node *temp)
	{
		//temp = head;
		if(temp == NULL)
		{
			return;
		}
		cout<<temp->data<<"\t";
		displayRecursion(temp->next);
	}

	void displayReverseRecursion(node *temp)
	{
		//temp = head;
		if(temp == NULL)
		{
			return;
		}
		displayReverseRecursion(temp->next);
		cout<<temp->data<<"\t";
	}

	void display()
	{
		node* temp = head;
		if(head==NULL)
			cout<<"Empty List!"<<endl;
		else
		{
			while(temp!=NULL)
			{
				cout<<temp->data<<"\t";
				temp=temp->next;
			}
			cout<<endl;
		}		
	}	

	void sortList(node* root)//Bubble sort
	{
		int flag;
		node* ptr = NULL;
		node* ptr1 = root;
		do
		{
			flag=0;
			ptr1 = root;
			while(ptr1->next!=ptr)
			{
				if(ptr1->data>ptr1->next->data)
				{
					int temp = ptr1->data;
					ptr1->data = ptr1->next->data;
					ptr1->next->data = temp;
					flag=1;
				}
				ptr1 = ptr1->next;
			}
			ptr = ptr1;
		}while(flag);
	}			
	
};

int main()
{
	node n1;
	
	cout<<"Enter the number of items you want to insert"<<endl;
	int n;
	cin>>n;
	for(int i=0;i<n;i++)
	{
		cout<<"Enter the data"<<endl;
		int d;
		cin>>d;
		n1.insertEnd(d);
	}
	node *temp = n1.getHead();
	cout<<"The elements of the linked list are"<<endl;
	n1.display();
	//n1.displayRecursion(temp);
	//cout<<endl;
	//n1.displayReverseRecursion(temp);
	//cout<<endl;
	//n1.insertPos(7,4);
	//n1.deleteNode(4);
	//n1.reverseList();
	//n1.reverseRecursion(temp);
	//n1.display();
	cout<<"List after bubble sorting:\n";
	n1.sortList(temp);
	n1.display();
	
	return 0;
}

File: Java/List/LinkedList.java
import java.util.Scanner;

class Node {
    private int data;
    private Node next;

    Node() {
        this.next = null;
    }

    Node(int data) {
        this.data = data;
        this.next = null;
    }

    public int getData() {
        return data;
    }

    public Node getNext() {
        return next;
    }

    public void setData(int data) {
        this.data = data;
    }

    public void setNext(Node next) {
        this.next = next;
    }
}

public class LinkedList {

    private Node head;
    private int length;

    LinkedList() {
        head = null;
    }

    public Node getHead() {
        return head;
    }

    private void insertFront(int val) {
        if (head == null) {
            head = new Node(val);
            length++;
            return;
        }
        Node n = new Node(val);
        n.setNext(head);
        head = n;
        length++;
    }

    private void insertEnd(int val) {
        if (head == null) {
            head = new Node(val);
            length++;
            return;
        }
        Node n = head;
        while (n.getNext() != null) {
            n = n.getNext();
        }
        n.setNext(new Node(val));
        length++;
    }

    private void insertPosition(int val, int position) {
        if (position > length || position < 0) {
            System.out.println("Specified position out of range: position=" + position);
            return;
        }
        if (position == 0) {
            Node node = new Node(val);
            node.setNext(head);
            head = node;
            length++;
            return;
        }
        Node n = head;
        int count = 0;
        while (count < position - 1) {
            n = n.getNext();
            count++;
        }
        Node node = new Node(val);
        node.setNext(n.getNext());
        n.setNext(node);
        length++;
    }

    private void deleteNode(int position) {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        if (position >= length || position < 0) {
            System.out.println("Specified position out of range: position=" + position);
            return;
        }
        if (position == 0) {
            head = head.getNext();
            return;
        }
        int count = 0;
        Node n = head;
        while (count < position - 1) {
            n = n.getNext();
            count++;
        }
        n.setNext(n.getNext().getNext());
    }

    private void printList() {
        Node n = head;
        while (n != null) {
            System.out.print(n.getData() + " ");
            n = n.getNext();
        }
        System.out.println();
    }

    private void printListRecursive(Node n) {
        if (n == null) {
            return;
        }
        System.out.print(n.getData() + " ");
        printListRecursive(n.getNext());
    }

    private void printListReverseRecursive(Node n) {
        if (n == null) {
            return;
        }
        printListReverseRecursive(n.getNext());
        System.out.print(n.getData() + " ");
    }

    private void reverseList() {
        Node previous = null;
        Node current = head;
        Node next = null;
        while (current != null) {
            next = current.getNext();
            current.setNext(previous);
            previous = current;
            current = next;
        }
        head = previous;
    }

    private void reverseListRecursive(Node n) {
        if (n.getNext() == null) {
            head = n;
            return;
        }
        reverseListRecursive(n.getNext());
        Node node = n.getNext();
        node.setNext(n);
        n.setNext(null);
    }

    public static void main(String args[]) {
        LinkedList list = new LinkedList();
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the number of items you want to insert: ");
        int n = sc.nextInt();
        for (int i = 0; i < n; i++) {
            // list.insertFront(sc.nextInt());
            list.insertEnd(sc.nextInt());
        }
        System.out.println("List Items: ");
        list.printList();

        // System.out.println("Enter the item and the position at which the item has to
        // be inserted: ");
        // list.insertPosition(sc.nextInt(), sc.nextInt());

        // System.out.println("Enter the position from the element has to be deleted:
        // ");
        // list.deleteNode(sc.nextInt());

        // System.out.println("List Items Recursive: ");
        // list.printListRecursive(list.getHead());
        // System.out.println();

        // System.out.println("List Items in Reverse Order: ");
        // list.printListReverseRecursive(list.getHead());
        // System.out.println();

        // list.reverseList();
        list.reverseListRecursive(list.getHead());
        System.out.println("List after Reversing: ");
        list.printList();

        sc.close();
    }
}
